File: python/ArithmeticSlicesIISubsequence.py
from typing import List


class Solution:
    def numberOfArithmeticSlices(self, A: List[int]) -> int:
        sum = 0
        dp = [{} for _ in A]  # 生成A长度的list内容是字典
        for i in range(len(A)):
            for j in range(i):
                diff = A[i]-A[j]
                dp[i][diff] = dp[i].get(diff, 0)+1
                if diff in dp[j]:
                    sum += dp[j][diff]
                    dp[i][diff] += dp[j][diff]
        return sum

File: Cplus/ArithmeticSlicesII-Subsequence.cpp
#include <climits>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution
{
public:
	int numberOfArithmeticSlices(vector<int> &A)
	{
		int N = A.size(), res = 0;
		vector<unordered_map<long, int>> dp(N);
		for (int i = 1; i < N; ++i)
		{
			for (int j = 0; j < i; ++j)
			{
				long diff = (long)A[i] - A[j];
				if (diff < INT_MIN || diff > INT_MAX)
					continue;
				auto iter = dp[j].find(diff);
				int count = iter == dp[j].end() ? 0 : iter->second;
				dp[i][diff] += count + 1;
				res += count;
			}
		}
		return res;
	}
};